File: src/main/scala/net/sf/xmlunit/builder/Input.scala
/*
  This file is licensed to You under the Apache License, Version 2.0
  (the "License"); you may not use this file except in compliance with
  the License.  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
*/
package net.sf.xmlunit.builder

import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.File
import java.io.InputStream
import java.io.IOException
import java.io.Reader
import java.io.StringReader
import java.net.MalformedURLException
import java.net.URI
import java.net.URISyntaxException
import java.net.URL
import javax.xml.transform.Source
import javax.xml.transform.TransformerFactory
import javax.xml.transform.URIResolver
import javax.xml.transform.dom.DOMSource
import javax.xml.transform.stream.StreamSource
import net.sf.xmlunit.exceptions.XMLUnitException
import org.w3c.dom.Document
import org.w3c.dom.Node

/**
 * Fluent API to create Source instances.
 */
object Input {
  trait Builder {
    /**
     * build the actual Source instance.
     */
    def build: Source
  }

  private class DOMBuilder(node: Node) extends Builder {
    val build = new DOMSource(node)
  }

  /**
   * Build a Source from a DOM Document.
   */
  def fromDocument(document: Document): Builder = new DOMBuilder(document)

  /**
   * Build a Source from a DOM Node.
   */
  def fromNode(node: Node): Builder = new DOMBuilder(node)

  private case class StreamBuilder(build: Source) extends Builder {
    def this(file: File) = this(new StreamSource(file))
    def this(stream: InputStream) = this(new StreamSource(stream))
    def this(reader: Reader) = this(new StreamSource(reader))
    
    def setSystemId(id: String) {
      Option(id).foreach(this.build.setSystemId(_))
    }
  }

  /**
   * Build a Source from a file.
   */
  def fromFile(file: File): Builder = new StreamBuilder(file)

  /**
   * Build a Source from a named file.
   */
  def fromFile(path: String): Builder = new StreamBuilder(new File(path))

  /**
   * Build a Source from a stream.
   */
  def fromStream(stream: InputStream): Builder = new StreamBuilder(stream)

  /**
   * Build a Source from a reader.
   */
  def fromReader(reader: Reader): Builder = new StreamBuilder(reader)

  /**
   * Build a Source from a string.
   */
  def fromMemory(s: String): Builder = this.fromReader(new StringReader(s))

  /**
   * Build a Source from an array of bytes.
   */
  def fromMemory(bytes: Array[Byte]): Builder = this.streamFromMemory(bytes)

  private def streamFromMemory(bytes: Array[Byte]): StreamBuilder =
    new StreamBuilder(new ByteArrayInputStream(bytes))

  /**
   * Build a Source from an URL.
   */
  def fromURL(url: URL): Builder = {
    var in: InputStream = null
    try {
      in = url.openStream()
      val out = new ByteArrayOutputStream
      val buffer = Array.ofDim[Byte](4096)
      var read = in.read(buffer)
      while (read >= 0) {
        if (read > 0) out.write(buffer, 0, read)
        read = in.read(buffer)
      }
    
      val builder = this.streamFromMemory(out.toByteArray)
      try builder.setSystemId(url.toURI.toString) catch {
        // impossible - shouldn't have been an URL in the first place
        case e: URISyntaxException => builder.setSystemId(url.toString)
      }
      builder
    } catch {
      case e: IOException => throw new XMLUnitException(e)
    } finally {
      if (in != null) in.close()
    }
  }

  /**
   * Build a Source from an URI.
   * @param uri must represent a valid URL
   */
  def fromURI(uri: URI): Builder = try this.fromURL(uri.toURL) catch {
    case e: MalformedURLException =>
      throw new IllegalArgumentException("uri " + uri + " is not an URL", e)
  }

  /**
   * Build a Source from an URI.
   * @param uri must represent a valid URL
   */
  def fromURI(uri: String): Builder = try this.fromURI(new URI(uri)) catch {
    case e: URISyntaxException =>
      throw new IllegalArgumentException("uri " + uri + " is not an URI", e)
  }

  trait TransformationSourceBuilder
    extends TransformationBuilderBase[TransformationSourceBuilder]
    with Builder {
    /**
     * Sets the stylesheet to use.
     */
    def withStylesheet(builder: Builder): TransformationSourceBuilder
  }

  private class Transformation(source: Source)
    extends TransformationBuilder[TransformationSourceBuilder](source)
    with TransformationSourceBuilder {
    def withStylesheet(builder: Builder): TransformationSourceBuilder =
      this.withStylesheet(builder.build)
    def build = new DOMSource(this.helper.transformToDocument)
  }

  /**
   * Build a Source by XSLT transforming a different Source.
   */
  def byTransforming(source: Source): TransformationSourceBuilder =
    new Transformation(source)

  /**
   * Build a Source by XSLT transforming a different Source.
   */
  def byTransforming(builder: Builder): TransformationSourceBuilder =
    return this.byTransforming(builder.build)
}

File: xmlunit/src/main/net-core/builder/Input.cs
/*
  This file is licensed to You under the Apache License, Version 2.0
  (the "License"); you may not use this file except in compliance with
  the License.  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
*/
using System;
using System.IO;
using System.Xml;
using net.sf.xmlunit.exceptions;
using net.sf.xmlunit.input;

namespace net.sf.xmlunit.builder {
    /// <summary>
    /// Fluent API to create ISource instances.
    /// </summary>
    public sealed class Input {
        public interface IBuilder {
            /// <summary>
            /// build the actual ISource instance.
            /// </summary>
            ISource Build();
        }

        internal class DOMBuilder : IBuilder {
            private readonly ISource source;
            internal DOMBuilder(XmlNode d) {
                source = new DOMSource(d);
            }
            public ISource Build() {
                return source;
            }
        }

        /// <summary>
        /// Build an ISource from a DOM Document.
        /// </summary>
        public static IBuilder FromDocument(XmlDocument d) {
            return new DOMBuilder(d);
        }

        /// <summary>
        /// Build an ISource from a DOM Node.
        /// </summary>
        public static IBuilder FromNode(XmlNode n) {
            return new DOMBuilder(n);
        }

        internal class StreamBuilder : IBuilder {
            private readonly ISource source;
            internal StreamBuilder(string s) {
                source = new StreamSource(s);
            }
            internal StreamBuilder(Stream s) {
                source = new StreamSource(s);
            }
            internal StreamBuilder(TextReader r) {
                source = new StreamSource(r);
            }
            internal string SystemId {
                set {
                    source.SystemId = value ?? string.Empty;
                }
            }
            public ISource Build() {
                return source;
            }
        }

        /// <summary>
        /// Build an ISource from a named file.
        /// </summary>
        public static IBuilder FromFile(string name) {
            return new StreamBuilder(name);
        }

        /// <summary>
        /// Build an ISource from a stream.
        /// </summary>
        public static IBuilder FromStream(Stream s) {
            StreamBuilder b = new StreamBuilder(s);
            if (s is FileStream) {
                b.SystemId = new Uri(Path.GetFullPath((s as FileStream).Name))
                    .ToString();
            }
            return b;
        }

        /// <summary>
        /// Build an ISource from a reader.
        /// </summary>
        public static IBuilder FromReader(TextReader r) {
            StreamBuilder b = new StreamBuilder(r);
            StreamReader s = r as StreamReader;
            if (s != null && s.BaseStream is FileStream) {
                b.SystemId =
                    new Uri(Path.GetFullPath((s.BaseStream as FileStream).Name))
                    .ToString();
            }
            return b;
        }

        /// <summary>
        /// Build an ISource from a string.
        /// </summary>
        public static IBuilder FromMemory(string s) {
            return FromReader(new StringReader(s));
        }

        /// <summary>
        /// Build an ISource from an array of bytes.
        /// </summary>
        public static IBuilder FromMemory(byte[] b) {
            return FromStream(new MemoryStream(b));
        }

        /// <summary>
        /// Build an ISource from an URI.
        /// <param name="uri">must represent a valid URL</param>
        /// </summary>
        public static IBuilder FromURI(string uri) {
            return new StreamBuilder(uri);
        }

        /// <summary>
        /// Build an ISource from an URI.
        /// <param name="uri">must represent a valid URL</param>
        /// </summary>
        public static IBuilder FromURI(System.Uri uri) {
            return new StreamBuilder(uri.AbsoluteUri);
        }

        public interface ITransformationBuilder
            : ITransformationBuilderBase<ITransformationBuilder>, IBuilder {
            /// <summary>
            /// Sets the stylesheet to use.
            /// </summary>
            ITransformationBuilder WithStylesheet(IBuilder b);
        }

        internal class Transformation
            : AbstractTransformationBuilder<ITransformationBuilder>,
              ITransformationBuilder {

            internal Transformation(ISource s) : base(s) {
            }
            public ITransformationBuilder WithStylesheet(IBuilder b) {
                return WithStylesheet(b.Build());
            }
            public ISource Build() {
                return new DOMSource(Helper.TransformToDocument());
            }
        }

        /// <summary>
        /// Build an ISource by XSLT transforming a different ISource.
        /// </summary>
        public static ITransformationBuilder ByTransforming(ISource s) {
            return new Transformation(s);
        }
        /// <summary>
        /// Build an ISource by XSLT transforming a different ISource.
        /// </summary>
        public static ITransformationBuilder ByTransforming(IBuilder b) {
            return ByTransforming(b.Build());
        }
    }
}
